Makefile

```diff
@@ -169,44 +174,44 @@ build: setup go-build
 # Target: go build
 #-----------------------------------------------------------------------------
 
-.PHONY: go-build
-
 # gobuild script uses custom linker flag to set the variables.
 # Params: OUT VERSION_PKG SRC
 
-.PHONY: pilot
-pilot: depend
-	bin/gobuild.sh ${GO_TOP}/bin/pilot-discovery istio.io/istio/pilot/tools/version ./pilot/cmd/pilot-discovery
+PILOT_GO_BINS:=${ISTIO_BIN}/pilot-discovery ${ISTIO_BIN}/pilot-agent \
+               ${ISTIO_BIN}/istioctl ${ISTIO_BIN}/sidecar-initializer
+$(PILOT_GO_BINS): depend
+	bin/gobuild.sh ${ISTIO_BIN}/$(@F) istio.io/istio/pilot/tools/version ./pilot/cmd/$(@F)
 
-.PHONY: pilot-agent
-pilot-agent: depend
-	bin/gobuild.sh ${GO_TOP}/bin/pilot-agent istio.io/istio/pilot/tools/version ./pilot/cmd/pilot-agent
+MIXER_GO_BINS:=${ISTIO_BIN}/mixs ${ISTIO_BIN}/mixc
+$(MIXER_GO_BINS): depend
+	bin/gobuild.sh ${ISTIO_BIN}/$(@F) istio.io/istio/mixer/pkg/version ./mixer/cmd/$(@F)
 
-.PHONY: istioctl
-istioctl: depend
-	bin/gobuild.sh ${GO_TOP}/bin/istioctl istio.io/istio/pilot/tools/version ./pilot/cmd/istioctl
+${ISTIO_BIN}/servicegraph: depend
```

Comment:
random thought/question:

is it typical for makefiles to explicitly depend on "depend" - isn't that automatically handled (through makedepend) - or is that only working for good old c/c++ programs ?

Reply:
I mostly followed the existing trend that had been set, but I think it's generally desirable for make targets to be as self-sufficient as possible (i.e., you can run "make foo" from an essentially clean system and things just build successfully).

The two tradeoffs is that 1) it can make each rule have redundant boilerplate (which is often handled using some template scheme or !include'ed rules file), or 2) slow things down as those dependencies are checked (and then people usually put in mechanism to speed up or short-circuit the checks [e.g., order-only dependencies or have "make depend" short-circuit based on the existence or timestamp of a sentinel file that's generated/updated when "depend" has run successfully).

Other schemes rely on people to first run something else like configure/autoconf before they get a usable makefile, but I'm mostly aware of those being used to scan for and complain about missing dependencies rather than take action to resolve them.